bundles/org.openhab.voice.porcupineks/src/main/java/org/openhab/voice/porcupineks/internal/PorcupineKSService.java

```diff
@@ -0,0 +1,283 @@
+        if (locale.getLanguage().equals(Locale.GERMAN.getLanguage())) {
+            var dePath = Path.of(PORCUPINE_FOLDER, "porcupine_params_de.pv");
+            if (Files.exists(dePath)) {
+                modelPath = dePath.toString();
+            } else {
+                logger.warn(
+                        "You can provide a specific model for de language in {}, english language model will be used",
+                        PORCUPINE_FOLDER);
+            }
+        } else if (locale.getLanguage().equals(Locale.FRENCH.getLanguage())) {
+            var frPath = Path.of(PORCUPINE_FOLDER, "porcupine_params_fr.pv");
+            if (Files.exists(frPath)) {
+                modelPath = frPath.toString();
+            } else {
+                logger.warn(
+                        "You can provide a specific model for fr language in {}, english language model will be used",
+                        PORCUPINE_FOLDER);
+            }
+        } else if (locale.getLanguage().equals("es")) {
+            var esPath = Path.of(PORCUPINE_FOLDER, "porcupine_params_es.pv");
+            if (Files.exists(esPath)) {
+                modelPath = esPath.toString();
+            } else {
+                logger.warn(
+                        "You can provide a specific model for es language in {}, english language model will be used",
+                        PORCUPINE_FOLDER);
+            }
+        }
+        if (modelPath == null) {
+            modelPath = prepareLib(bundleContext, Porcupine.MODEL_PATH);
+        }
+        return modelPath;
+    }
+
+    private String getKeywordResourcePath(BundleContext bundleContext, String keyWord) throws IOException {
+        var localKeywordFile = keyWord.toLowerCase().replace(" ", "_") + ".ppn";
+        var localKeywordPath = Path.of(PORCUPINE_FOLDER, localKeywordFile);
+        if (Files.exists(localKeywordPath)) {
+            return localKeywordPath.toString();
+        }
+        String env = getPorcupineEnv();
+        var keywordPath = "porcupine/resources/keyword_files/" + env + "/" + keyWord.toLowerCase() + "_" + env + ".ppn";
+        return prepareLib(bundleContext, keywordPath);
+    }
+
+    private void processInBackground(Porcupine porcupine, KSListener ksListener, AudioStream audioStream) {
+        int numBytesRead;
+        // buffers for processing audio
+        int frameLength = porcupine.getFrameLength();
+        ByteBuffer captureBuffer = ByteBuffer.allocate(frameLength * 2);
+        captureBuffer.order(ByteOrder.LITTLE_ENDIAN);
+        short[] porcupineBuffer = new short[frameLength];
+        this.loop = true;
+        while (loop) {
+            try {
+                // read a buffer of audio
+                numBytesRead = audioStream.read(captureBuffer.array(), 0, captureBuffer.capacity());
+                // don't pass to porcupine if we don't have a full buffer
+                if (numBytesRead != frameLength * 2) {
+                    continue;
```

Comment:
Question: What if the read returned only part of the buffer ? You are ignoring it.

Reply:
That read is calling a native method so I think is always blocking so I maybe no need for the thread sleep, but I'm not totally sure.
Those line are almost an exactly copy of the porcupine java sample, I even left same comment. I can not see the binary implementation, I also think there is no change of read not returning the full buffer, as least on an infinite stream.

Reply:
Ok, fine.

Reply:
Finally I have added a sleep before the continue clause. Is this enough?